scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -636,200 +646,107 @@ def dijkstra(csgraph, directed=True, indices=None,
     else:
         return dist_matrix.reshape(return_shape)
 
-@cython.boundscheck(False)
-cdef _dijkstra_setup_heap_multi(FibonacciHeap *heap,
-                                FibonacciNode* nodes,
-                                const int[:] source_indices,
-                                int[:] sources,
-                                double[:] dist_matrix,
-                                int return_pred):
-    cdef:
-        unsigned int Nind = source_indices.shape[0]
-        unsigned int N = dist_matrix.shape[0]
-        unsigned int i, k, j_source
-        FibonacciNode *current_node
-
-    for k in range(N):
-        initialize_node(&nodes[k], k)
-
-    heap.min_node = NULL
-    for i in range(Nind):
-        j_source = source_indices[i]
-        current_node = &nodes[j_source]
-        if current_node.state == SCANNED:
-            continue
-        dist_matrix[j_source] = 0
-        if return_pred:
-            sources[j_source] = j_source
-        current_node.state = SCANNED
-        current_node.source = j_source
-        insert_node(heap, &nodes[j_source])
-
-@cython.boundscheck(False)
-cdef _dijkstra_scan_heap_multi(FibonacciHeap *heap,
-                               FibonacciNode *v,
-                               FibonacciNode* nodes,
-                               const double[:] csr_weights,
-                               const int[:] csr_indices,
-                               const int[:] csr_indptr,
-                               int[:] pred,
-                               int[:] sources,
-                               int return_pred,
-                               DTYPE_t limit):
-    cdef:
-        unsigned int j_current
-        ITYPE_t j
-        DTYPE_t next_val
-        FibonacciNode *current_node
 
-    for j in range(csr_indptr[v.index], csr_indptr[v.index + 1]):
-        j_current = csr_indices[j]
-        current_node = &nodes[j_current]
-        if current_node.state != SCANNED:
-            next_val = v.val + csr_weights[j]
-            if next_val <= limit:
-                if current_node.state == NOT_IN_HEAP:
-                    current_node.state = IN_HEAP
-                    current_node.val = next_val
-                    current_node.source = v.source
-                    insert_node(heap, current_node)
-                    if return_pred:
-                        pred[j_current] = v.index
-                        sources[j_current] = v.source
-                elif current_node.val > next_val:
-                    current_node.source = v.source
-                    decrease_val(heap, current_node,
-                                 next_val)
-                    if return_pred:
-                        pred[j_current] = v.index
-                        sources[j_current] = v.source
+ctypedef unsigned int uint_t
+ctypedef pair[DTYPE_t, uint_t] dist_index_pair_t
+ctypedef priority_queue[dist_index_pair_t] dijkstra_queue_t
```

Comment:
I didn't know that. That's a cool bit of knowledge. Thanks for sharing

Reply:
You are welcome. 🙂

This is definitely doable in C++, and it should also be in Cython.

I have checked that Cython exposes all the interfaces so that it is possible; if it is not, then one can import the missing bits manually.

Reply:
Great idea. Unfortunately, this is not supported in Cython.
See Cythonic `priotity_queue` [here](https://github.com/cython/cython/blob/master/Cython/Includes/libcpp/queue.pxd#L15).

Reply:
Indeed, it is not possible with the current state of the Cython interfaces to the C++ standard library.

Looking at [`std::unordered_map`](https://github.com/cython/cython/blob/master/Cython/Includes/libcpp/unordered_map.pxd#L4), it seems like it can be modified so that it is supported, but we would need to wait for the next version of Cython and make sure it is used.

You can ignore this suggestion.